fiftyone/brain/internal/core/leaky_splits.py

```diff
@@ -0,0 +1,383 @@
+                sim_conf_dict_kwargs["name"] = self.config.similarity_backend
+            if self.config.embeddings_field is not None:
+                sim_conf_dict_kwargs[
+                    "embeddings_field"
+                ] = self.config.embeddings_field
+            if self.config.model is not None:
+                sim_conf_dict_kwargs["model"] = self.config.model
+            if self.config.model_kwargs is not None:
+                sim_conf_dict_kwargs["model_kwargs"] = self.config.model_kwargs
+
+            # empty conf if conf dict isn't provided
+            similarity_conf_provided = {}
+            if self.config.similarity_config_dict is not None:
+                similarity_conf_provided = self.config.similarity_config_dict
+            sim_conf_dict_kwargs = {
+                "name": None,  # default if user doesn't provide any value
+                "model": _DEFAULT_MODEL,  # default if user doesn't provide any value
+                **similarity_conf_provided,  # conf over defaults
+                **sim_conf_dict_kwargs,  # arguments over conf
+            }
+
+            backend_name = sim_conf_dict_kwargs.pop("name")
+
+            similarity_config_object = sim._parse_config(
+                backend_name, **sim_conf_dict_kwargs
+            )
+            self._similarity_method = similarity_config_object.build()
+            self._similarity_method.ensure_requirements()
+
+            self._similarity_index = self._similarity_method.initialize(
+                samples, self.config.similarity_brain_key
+            )
+
+    def set_threshold(self, threshold):
+        """Set threshold for leak computation"""
+        self._leak_threshold = threshold
+
+    @property
+    def leaks(self):
+        """
+        Returns view with all potential leaks.
+        """
+
+        # access cache if possible
+        if self._last_computed_threshold == self._leak_threshold:
+            return self._leaks
+
+        # populate index if it doesn't have some samples
+        if not self._similarity_index.total_index_size == len(self.samples):
+            (
+                embeddings,
+                sample_ids,
+                label_ids,
+            ) = self._similarity_index.compute_embeddings(self.samples)
+            self._similarity_index.add_to_index(
+                embeddings, sample_ids, label_ids
+            )
+
+        # check if duplicates already computed in the index
+        if not self._similarity_index.thresh == self._leak_threshold:
+            self._similarity_index.find_duplicates(self._leak_threshold)
+        duplicates = self._similarity_index.duplicates_view()
+
+        # filter duplicates to just those with neighbors in different splits
+        to_keep = []
+        for (
+            sample_id,
+            neighbors,
+        ) in self._similarity_index.neighbors_map.items():
+            keep_sample = False
+            sample_split = self.id2split[sample_id]
```

Comment:
```py

view1 = dataset.limit(500)
view2 = dataset.skip(500).limit(500)

fob.compute_leaky_splits(dataset, split_views={"train":view1, "test":view2}, brain_key="leak_key2", embeddings_field="clip_embeddings")
```

Results in the following error:

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Cell In[23], [line 4](vscode-notebook-cell:?execution_count=23&line=4)
      [1](vscode-notebook-cell:?execution_count=23&line=1) view1 = dataset.limit(500)
      [2](vscode-notebook-cell:?execution_count=23&line=2) view2 = dataset.skip(500).limit(500)
----> [4](vscode-notebook-cell:?execution_count=23&line=4) fob.compute_leaky_splits(dataset, split_views={"train":view1, "test":view2}, brain_key="leak_key2", embeddings_field="clip_embeddings")

File ~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:787, in compute_leaky_splits(samples, brain_key, split_views, split_field, split_tags, threshold, similarity_brain_key, embeddings_field, model, model_kwargs, similarity_backend, similarity_config_dict, **kwargs)
    [723](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:723) """Uses a similarity index or creates one on the spot to find leaks.
    [724](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:724) 
    [725](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:725) Calling this method only creates the index. You can then call the methods
   (...)
    [782](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:782)     a :class:`fiftyone.core.view.DatasetView`
    [783](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:783) """
    [785](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:785) from fiftyone.brain.internal.core.leaky_splits import compute_leaky_splits
--> [787](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:787) return compute_leaky_splits(
    [788](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:788)     samples,
    [789](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:789)     brain_key=brain_key,
    [790](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:790)     split_views=split_views,
    [791](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:791)     split_field=split_field,
    [792](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:792)     split_tags=split_tags,
    [793](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:793)     threshold=threshold,
    [794](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:794)     similarity_brain_key=similarity_brain_key,
    [795](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:795)     embeddings_field=embeddings_field,
    [796](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:796)     model=model,
    [797](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:797)     model_kwargs=model_kwargs,
    [798](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:798)     similarity_backend=similarity_backend,
    [799](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:799)     similarity_config_dict=similarity_config_dict,
    [800](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:800)     **kwargs,
    [801](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/__init__.py:801) )

File ~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:67, in compute_leaky_splits(samples, brain_key, split_views, split_field, split_tags, threshold, similarity_brain_key, embeddings_field, model, model_kwargs, similarity_backend, similarity_config_dict, **kwargs)
     [61](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:61) if results._save_similarity_index:
     [62](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:62)     results._similarity_method.register_run(
     [63](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:63)         samples,
     [64](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:64)         results.config.similarity_brain_key,
     [65](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:65)     )
---> [67](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:67) results.set_threshold(threshold)
     [68](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:68) leaks = results.leaks
     [70](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:70) if brain_key is not None:

File ~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:231, in leaks(self)
    [226](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:226) to_keep = []
    [227](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:227) for (
    [228](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:228)     sample_id,
    [229](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:229)     neighbors,
    [230](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:230) ) in self._similarity_index.neighbors_map.items():
--> [231](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:231)     keep_sample = False
    [232](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:232)     sample_split = self.id2split[sample_id]
    [233](https://file+.vscode-resource.vscode-cdn.net/Users/jacobmarks/Desktop/~/Desktop/work/brain/fiftyone-brain/fiftyone/brain/internal/core/leaky_splits.py:233)     leaks = []

KeyError: '673f6702a7ac3f16db0b8c32'
```

After making the following progress:
```plaintext
Computing embeddings...
 100% |███████████████| 5000/5000 [5.1m elapsed, 0s remaining, 17.4 samples/s]      
Computing duplicate samples...
Generating index for 5000 embeddings...
Index complete
Generating neighbors graph for 4765 embeddings...
Index complete
Duplicates computation complete
```

Reply:
This happens because self.samples has more samples than the union of the splits. The id2split dict is constructed on the splits, but the index is constructed on all of the self.samples. As we discussed above, the check to see if the union of the splits and self.samples is the same is not scalable. I can add another heuristic check (i.e. size), but I don't really like those. Honestly I'm fine wrapping id2split in a try except and if it fails throwing an error with something along the lines of "KeyError ... This usually happens when the samples, split, and similarity index don't line up. Please make sure the samples are the union of the splits and that if a similarity index is provided it was applied on a superset of samples". What do you guys think @jacobmarks @brimoor